structured/privilege.go

```diff
@@ -0,0 +1,258 @@
+		pv[i] = Privilege(v)
+		i++
+	}
+	sort.Sort(pv)
+	return pv
+}()
+
+// String implements the Stringer interface.
+// This keeps the existing order and uses ", " as separator.
+func (pl PrivilegeList) String() string {
+	ret := make([]string, len(pl), len(pl))
+	for i, p := range pl {
+		ret[i] = Privilege_name[int32(p)]
+	}
+	return strings.Join(ret, ", ")
+}
+
+// SortedString is similar to String() but returns
+// privileges sorted by name and uses "," as separator.
+func (pl PrivilegeList) SortedString() string {
+	ret := make([]string, len(pl), len(pl))
+	for i, p := range pl {
+		ret[i] = Privilege_name[int32(p)]
+	}
+	sort.Strings(ret)
+	return strings.Join(ret, ",")
+}
+
+// rawToPrivilegeList converts a set of privileges
+// into a PrivilegeList.
+func rawToPrivilegeList(m uint32) PrivilegeList {
+	ret := PrivilegeList{}
+	for _, p := range privilegeValues {
+		if m&(1<<uint32(p)) != 0 {
+			ret = append(ret, Privilege(p))
+		}
+	}
+
+	return ret
+}
+
+// Sort methods for the PrivilegeDescriptor.Users list.
+type userPrivilegeList []*UserPrivileges
+
+func (upl userPrivilegeList) Len() int {
+	return len(upl)
+}
+
+func (upl userPrivilegeList) Swap(i, j int) {
+	upl[i], upl[j] = upl[j], upl[i]
+}
+
+func (upl userPrivilegeList) Less(i, j int) bool {
+	return upl[i].User < upl[j].User
+}
+
+// findUser looks for a specific user in the list.
+// Returns (nil, false) if not found, or (obj, true) if found.
+func (p *PrivilegeDescriptor) findUser(user string) (*UserPrivileges, bool) {
+	idx := sort.Search(len(p.Users), func(i int) bool {
+		return p.Users[i].User >= user
+	})
+	if idx < len(p.Users) && p.Users[idx].User == user {
+		return p.Users[idx], true
+	}
+	return nil, false
+}
+
+// findOrCreateUser looks for a specific user in the list, creating it if needed.
+func (p *PrivilegeDescriptor) findOrCreateUser(user string) *UserPrivileges {
+	idx := sort.Search(len(p.Users), func(i int) bool {
+		return p.Users[i].User >= user
+	})
+	if idx == len(p.Users) {
+		// Not found but should be inserted at the end.
+		p.Users = append(p.Users, &UserPrivileges{User: user})
+	} else if p.Users[idx].User == user {
+		// Found.
+	} else {
+		// New element to be inserted at i.
+		p.Users = append(p.Users, nil)
+		copy(p.Users[idx+1:], p.Users[idx:])
+		p.Users[idx] = &UserPrivileges{User: user}
+	}
+	return p.Users[idx]
+}
+
+// removeUser looks for a given user in the list and removes it if present.
+func (p *PrivilegeDescriptor) removeUser(user string) {
+	idx := sort.Search(len(p.Users), func(i int) bool {
+		return p.Users[i].User >= user
+	})
+	if idx == len(p.Users) || p.Users[idx].User != user {
+		// Not found.
+		return
+	}
+
+	copy(p.Users[idx:], p.Users[idx+1:])
+	p.Users[len(p.Users)-1] = nil
+	p.Users = p.Users[:len(p.Users)-1]
+}
+
+// DefaultDatabasePrivilegeDescriptor returns a privilege descriptor
+// with ALL privileges for the root user.
+func DefaultDatabasePrivilegeDescriptor() *PrivilegeDescriptor {
```

Comment:
Can this be a variable instead of a function? I realize making it a function prevents someone from changing the variable, but nobody will be naughty enough to do that.